test/distributed/test_distributed.py

```diff
@@ -2678,6 +2688,251 @@ def test_nccl_gather_object_err(self):
+    @require_backends_available({"gloo", "nccl"})
+    @skip_if_lt_x_gpu(2)
+    @skip_if_rocm
+    def test_ddp_grad_div_uneven_inputs(self):
+        # Test that we scale by the effective world size when allreducing grads.
+        # For example if N processes start DDP training but 0 < K < N join, we
+        # should divide by N - K and not N.
+        dim = 5
+        batch = 1
+        grad_scale = 50
+        rank = self.rank
+        model = nn.Linear(dim, dim, bias=False)
+        inp = torch.ones(batch, dim).to(self.rank) * grad_scale
+        net = torch.nn.parallel.DistributedDataParallel(
+            model.cuda(rank), device_ids=[self.rank], bucket_cap_mb=1
+        )
+        n_iters = 1 if self.rank == 0 else 2
+        from torch.nn.parallel.distributed import join
+
+        with join(net) as ddp_join:
+            for i in range(n_iters):
+                loss = net(inp).sum()
+                loss.backward()
+                # The grad is always expected_grad, since we divide by the number
+                # of currently active processes and inactive processes contribute
+                # zero gradient. If we kept dividing by static initial world
+                # size as processes leave, the grad would be smaller.
+                expected_grad = torch.ones(dim, dim).to(self.rank) * grad_scale
+                param = list(net.parameters())[0]
+                self.assertEqual(expected_grad, param.grad)
+                # Avoid accumulating grads so that it's the same every iteration
+                net.zero_grad()
+                torch.cuda.synchronize(device=self.rank)
+
+    def _run_uneven_inputs_test(
+        self, test_case, num_iters, iteration_offset, find_unused_params
+    ):
+        from torch.nn.parallel.distributed import join
+
+        model = test_case.model
+        inp = test_case.inp
+        rank = self.rank
+        # Ensure all outsanding GPU work is comlete so this test runs independently.
+        torch.cuda.synchronize()
+        # Bucket_cap_mb is intentionally low to test allreduce scheduling when
+        # there are many buckets.
+        net = torch.nn.parallel.DistributedDataParallel(
+            model.cuda(rank),
+            device_ids=[rank],
+            bucket_cap_mb=1,
+            find_unused_parameters=find_unused_params,
+        )
+        if rank != 0:
+            num_iters += iteration_offset
+
+        with join(net) as ddp_join:
+            for _ in range(num_iters):
+                if isinstance(inp, tuple):
+                    loss = net(*inp).sum()
+                else:
+                    loss = net(inp).sum()
+                loss.backward()
+                self._model_step(net)
+                # Ensure completion of GPU kernels (including allreduce). If the
+                # join API is not properly implemented, then this should hang
+                # since the allreduce will hang.
+                torch.cuda.synchronize(device=rank)
```

Comment:
For testing purposes, I figured that having a sync every iteration would be useful since the test would fail (hang) immediately, rather than trying to do additional iterations, and it may give us a closer view into where the error occurred.